static/src/javascripts/projects/common/modules/spacefinder.js

```diff
@@ -196,6 +204,76 @@ const enforceRules = (measurements, rules, exclusions) => {
 	return candidates;
 };
 
+const colours = {
+	red: 'rgb(255 178 178)',
+	orange: 'rgb(255 213 178)',
+	yellow: 'rgb(254 255 178)',
+	blue: 'rgb(178 248 255)',
+	pink: 'rgb(255 178 242)',
+	green: 'rgb(178 255 184)',
+};
+
+const addHoverListener = (candidate, tooClose) => {
+	tooClose.forEach((opponent) => {
+		candidate.addEventListener(
+			'mouseenter',
+			() =>
+				(opponent.style.cssText = `border: thick solid ${colours.blue}; color: ${colours.blue}`),
+		);
+
+		candidate.addEventListener(
+			'mouseleave',
+			() => (opponent.style.cssText = ''),
+		);
+	});
+};
+
+const markCandidates = (exclusions, winners, options) => {
```

Comment:
Have you considered defining spacefinder visualizer styles in a stylesheet and having `markCandidates` just be responsible for adding extra classes to candidates that the css would select for?
I think the advantages of that approach would be that we could keep the code in this file dedicated to debug mode very light and make the new styles we're introducing a bit more discoverable as they'd be alongside other styles.

Reply:
This is a good idea. Do you have any thoughts on where these styles would live? I was thinking in [here somewhere](https://github.com/guardian/dotcom-rendering/blob/main/dotcom-rendering/src/web/components/TextBlockComponent.tsx) but I couldn't get it to work.

I'm new to the concept of CSS-in-JS, it's blowing my mind a bit!

Reply:
Hmm good question. Seeing as how the rest of the article styles live in DCR it might actually be more trouble than it's worth to try to colocate them.
@ashishpuliyel and I were just discussing this and his suggestion was to have the js insert a style element with spacefinder debug styles only if we're in `sfdebug` mode.
A drawback of this approach as well as manipulating elements' `style.cssText` is that our css will live in strings where we can't benefit from syntax highlighting, validation, etc.

Reply:
On one hand, given DCR renders 100% of ALL article content (emphasis because it _just_ happened 😎 ), it makes sense to have the styles on that platform but if the spacefinder code is here then maybe keeping them co-located makes more sense?

Reply:
Thanks for weighing in with the dotcom perspective @oliverlloyd. 
`sfdebug` is being read from the url parameters.
I can also see both the case for keeping the spacefinder debugging styles as close to spacefinder as possible and for keeping them with the other article styles. An argument for the former approach is the fact that the debugger is a purely internal tool and so might not need to be as optimised as the rest of our article-rendering code?
Another consideration is that we'd like to extend spacefinder debug mode to liveblog adverts which aren't yet rendered by DCR (if I understand right).

Reply:
My vote is for injecting the styles via JS - as it is now.
It keeps us platform independent and we don't have to worry about adding external styling.
That might change in future if the CSS gets more complicated but for now imo we can keep it to an MVP.

A small improvement might be moving the styling injection to a separate module to keep it away from the JS?

Reply:
Sounds good @arelra, is [this](https://github.com/guardian/frontend/pull/24618/commits/e1b091be973d32f408a0e1a8650a6605430d6e81) what you mean?